build_resolvers/lib/src/analysis_driver_model.dart

```diff
@@ -82,66 +90,84 @@ class AnalysisDriverModel {
-        /// modified rather than added, fix the implementation here.
-        resourceProvider.newFile(id.asPath, content);
-      } else {
-        if (resourceProvider.getFile(id.asPath).exists) {
-          resourceProvider.deleteFile(id.asPath);
-        }
-      }
-    }
-
-    // Notify the analyzer of changes.
+    // Immediately take the lock on `driver` so that the whole class state,
+    // `_graph` and `_readForAnalyzir`, is only mutated by one build step at a
+    // time. Otherwise, interleaved access complicates processing significantly.
     await withDriverResource((driver) async {
-      for (final id in ids) {
-        // TODO(davidmorgan): add test coverage for over-notification of
-        // changes, fix the implementaion here.
-        driver.changeFile(id.asPath);
-      }
-      await driver.applyPendingFileChanges();
+      return _performResolve(driver, buildStep, entryPoints, withDriverResource,
+          transitive: transitive);
     });
   }
 
-  /// Walks the import graph from [ids], returns full transitive deps.
-  Future<Set<AssetId>> _expandToTransitive(
-      AssetReader reader, Iterable<AssetId> ids) async {
-    final result = ids.toSet();
-    final nextIds = Queue.of(ids);
-    while (nextIds.isNotEmpty) {
-      final nextId = nextIds.removeFirst();
+  Future<void> _performResolve(
+      AnalysisDriverForPackageBuild driver,
+      BuildStep buildStep,
+      List<AssetId> entryPoints,
+      Future<void> Function(
+              FutureOr<void> Function(AnalysisDriverForPackageBuild))
+          withDriverResource,
+      {required bool transitive}) async {
+    var idsToSyncOntoResourceProvider = entryPoints;
+    Iterable<AssetId> inputIds = entryPoints;
 
-      // Skip if not readable. Note that calling `canRead` still makes it a
-      // dependency of the `BuildStep`.
-      if (!await reader.canRead(nextId)) continue;
+    // If requested, find transitive imports.
+    if (transitive) {
+      await _graph.load(buildStep, entryPoints);
+      idsToSyncOntoResourceProvider = _graph.nodes.keys.toList();
+      inputIds = _graph.inputsFor(entryPoints);
 
-      final content = await reader.readAsString(nextId);
-      final deps = _parseDependencies(content, nextId);
+      // Check for missing inputs that were written during the build.
+      for (final id in inputIds
+          .where((id) => !id.path.endsWith(_transitiveDigestExtension))) {
+        if (_graph.nodes[id]!.isMissing) {
+          if (await buildStep.canRead(id)) {
+            idsToSyncOntoResourceProvider.add(id);
+            _syncedOntoResourceProvider.remove(id);
+          }
+        }
+      }
+    }
 
-      // For each dep, if it's not in `result` yet, it's newly-discovered:
-      // add it to `nextIds`.
-      for (final dep in deps) {
-        if (result.add(dep)) {
-          nextIds.add(dep);
+    // Notify [buildStep] of its inputs.
+    for (final id in inputIds) {
+      await buildStep.canRead(id);
```

Comment:
Does "canRead" say "this is what you should process" or does it say "this is what you are allowed to read" or how am I supposed to read the code in combination with the comment?

Reply:
Also possibly some were already "canRead" above (in `if (transitive)`). Maybe this could be something like

```dart
if (transitive) {
  // bla bla
  for (final id in inputIds) {
    // Notify [buildStep] of its inputs. And also other stuff.
    if (await buildStep.canRead(id) && !id.path.endsWith(_transitiveDigestExtension) && _graph.nodes[id]!.isMissing) {
      // Some good comment here.
      idsToSyncOntoResourceProvider.add(id);
      _syncedOntoResourceProvider.remove(id);
    }
  }
} else {
  // Notify [buildStep] of its inputs.
  for (final id in inputIds) {
    await buildStep.canRead(id);
  }
}

Reply:
`canRead` asks the reader if the file is readable (exists + is allowed to be read), _and_ marks the file as an input to the `buildStep`. I need to untangle those two soon to make progress.

 - notifying about inputs needs to be separate from reading; we can't efficiently track inputs while passing them around one by one
 - there should be no need to eagerly check whether a file was written, `build_runner` can track which files get written
 
so I'll leave it separate as after those changes the reading and notifying will actually be separate :)

Thanks.